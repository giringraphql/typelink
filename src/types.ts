export type ResolvedValue<T> = T | Promise<T>;
export type ResolvedList<T> = T[] | Promise<T[]> | Promise<T>[];

export interface ConcreteClass<T = any> {
  new(...args: any[]): T;
}

export type Instantiator<TClass = any> = (value: { [key: string]: any }) => TClass;

export function defaultInputFieldInstantiator(value: any) {
  return value;
}

export type Thunk<T> = () => T;
