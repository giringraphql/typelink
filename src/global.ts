import { loadFallbackRootTypes } from './definition/rootTypes';
import { loadBuiltInScalar } from './definition/scalar';
import { DefinitionKind } from './metadata/Definition';
import { MetadataStorage } from './metadata/MetadataStorage';
import { DefinitionParser } from './sdl/ast';
import { coerceType } from './type-expression/coerceType';
import { TypeExpression } from './type-expression/TypeExpression';
import { Thunk } from './types';
import { TypeArg } from './type-expression/types';


/**
 * Create a new MetadataStorage initialized with default metadata
 */
export function createMetadataStorage() {
  const storage = new MetadataStorage();
  loadBuiltInScalar(storage);
  loadFallbackRootTypes(storage);
  return storage;
}

/**
 * Global MetadataStorage used by default.
 */
export let globalMetadataStorage: MetadataStorage;

export function getGlobalMetadataStorage() {
  if (!globalMetadataStorage) {
    globalMetadataStorage = createMetadataStorage();
  }
  return globalMetadataStorage;
}

/**
 * Get a GraphQLType instance from the given storage or default
 * global metadata storage.
 * @param typeArg
 * @param storage
 */
export function getType(typeArg: TypeExpression | TypeArg, kind: DefinitionKind = 'any'): any {
  const storage = getGlobalMetadataStorage();
  return coerceType(typeArg).getType({ kind, storage });
}

/**
 * Define a type linked to decorated class and add it to the given
 * storage or default global metadata storage.
 * @param metadataOrThunk
 */
export function defineType(parsersOrThunk: DefinitionParser[] | Thunk<DefinitionParser[]>) {
  const storage = getGlobalMetadataStorage();
  return function defDecoratorFn<T extends Function>(definitionClass: T) {
    if (Array.isArray(parsersOrThunk)) {
      parsersOrThunk.forEach(parser => {
        storage.registerMetadata(parser.parse(definitionClass));
      });
    } else {
      storage.deferRegister(() => {
        parsersOrThunk().forEach(parser => {
          storage.registerMetadata(parser.parse(definitionClass));
        });
      });
    }
    return definitionClass;
  };
}
