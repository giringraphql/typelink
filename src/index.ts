export { gql } from './sdl/gql';

export * from './definition/GraphQLTypeIndex';
export * from './definition/InputType';
export * from './definition/InterfaceType';
export * from './definition/ObjectType';
export * from './definition/rootTypes';
export * from './definition/scalar';
export * from './definition/SubscriptionType';


export * from './reference/Field';
export * from './reference/Implement';
export * from './reference/InputField';
export * from './global';

export * from './type-expression/DefinitionTypeExpression';
export * from './type-expression/GraphQLTypeExpression';
export * from './type-expression/TypeExpression';
export * from './type-expression/structure';
export * from './type-expression/coerceType';
export * from './type-expression/types';

export * from './metadata/Definition';
export * from './metadata/MetadataStorage';
export * from './metadata/Reference';

export * from './relay/connection';
export * from './relay/mutation';
export * from './relay/node';
