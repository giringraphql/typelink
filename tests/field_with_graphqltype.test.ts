import {
  GraphQLObjectType,
  GraphQLString,
  GraphQLSchema,
  graphql,
  printSchema,
  GraphQLEnumType
} from 'graphql';

import { getType, gql, defineType } from '../src';


@defineType(gql`
  type ItemInfo {
    description: String
  }
`)
class ItemInfo {
  description() {
    return 'foobarbaz';
  }
}

const itemType = new GraphQLObjectType({
  name: 'Item',
  fields() {
    return {
      name: {
        type: GraphQLString,
        resolve() {
          return 'foo';
        }
      },
      info: {
        type: getType(ItemInfo),
        resolve() { return new ItemInfo(); }
      }
    };
  },
});

const bazEnum = new GraphQLEnumType({
  name: 'Baz',
  values: { A: {}, B: {}, C: {}, }
});

@defineType(gql`
  type Query {
    item: ${itemType}
    nonNullItem: ${itemType}!
    baz: ${bazEnum}
    erroneousBaz: ${bazEnum}
  }
`)
class Query {
  static item() { return {}; }
  static nonNullItem() { return {}; }
  static baz(): 'A' | 'B' | 'C' {
    return 'A';
  }
  static erroneousBaz(): 'A' | 'B' | 'C' {
    return 'D' as any;
  }
}

const schema = new GraphQLSchema({
  query: getType(Query)
});

describe('field with GraphQLType', () => {
  test('generates schema as expected', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  test('query', async () => {
    const result = await graphql({ schema, source: `
      query {
        item {
          name
          info { description }
        }
        nonNullItem {
          name
          info { description }
        }
      }
    `});
    expect(result.data!.item.name).toEqual('foo');
    expect(result.data!.item.info.description).toEqual('foobarbaz');
    expect(result.data!.nonNullItem.name).toEqual('foo');
    expect(result.data!.nonNullItem.info.description).toEqual('foobarbaz');
  });

  test('query enum', async () => {
    let result: any;

    result = await graphql({ schema, source: `
      query {
        baz
      }
    `});
    expect(result.data!.baz).toEqual('A');

    result = await graphql({ schema, source: `
      query {
        erroneousBaz
      }
    `});
    expect(result.errors[0].message).toEqual('Expected a value of type "Baz" but received: "D"');
  });
});
