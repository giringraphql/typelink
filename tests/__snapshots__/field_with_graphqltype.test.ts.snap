// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field with GraphQLType generates schema as expected 1`] = `
"enum Baz {
  A
  B
  C
}

type Item {
  name: String
  info: ItemInfo
}

type ItemInfo {
  description: String
}

type Query {
  item: Item
  nonNullItem: Item!
  baz: Baz
  erroneousBaz: Baz
}
"
`;
